Version 0.1

piyanatk opened the pull request:
This PR contains significant reformatting of the Python codes to compile with [PEP8](https://www.python.org/dev/peps/pep-0008/), improvement on the packaging, and the use of `setuptools-scm` to automate versioning.

### Code Styling
- `sparse_2level` was renamed to `Sparse2Level` to compile with PEP8 (i.e., class must use CamelCase).
- `np` namespace is used instead of `numpy`.
- Various PEP8 formatting.

### Packaging
- `ctypes` wrapped functions were migrated from the main Python code, `corrcal.py`, to `wrapper.py`. This separate the low-level, `ctypes` wrapped, functions from the high-level Python interface and should make them easier to maintain in a long run.

### Versioning
- `setuptools-scm` was implemented, as suggested by @steven-murray, to automate code versioning based on the `git` hash. See [https://pypi.org/project/setuptools-scm/#description](https://pypi.org/project/setuptools-scm/#description).

### Other
- The `do_timing` code blocks in `corrcal.py` seem to be a hack to time the codes. They are probably not being used, and there are better way to do profiling. Therefore, we commented out these lines for now and will removed them in the future.

One thing that has not been fixed is duplicated code fragmentation. There are many duplicated blocks of codes in `corrcal.py`. These lines are redundant and prone to bugs. Fixing them will require thorough understanding of how the codes work, so it will take some time.

piyanatk committed:
Reformatting to PEP8 and commenting out do_time lines

piyanatk committed:
PEP8 reformatting

piyanatk committed:
PEP8 reformatting

piyanatk committed:
Merge remote-tracking branch 'origin/restructure' into restructure

# Conflicts:
#	corrcal/corrcal.py

piyanatk committed:
Rename

piyanatk committed:
Move `ctypes` library link to `wrapper.py`

piyanatk committed:
First commit.

Migrate `ctypes` library link from to `corrcal.py` to this file.

piyanatk committed:
Rename header file to match previous commit.

piyanatk committed:
- Rename output c library to `c_corrcal`
- Update version to 0.0.2
- Add `numpy` requirement

piyanatk commented:
New commits:
- fa66806, fb43cfe, 0afd1bc, 33290bb, and 1d0c174 migrate `ctypes` library wrapping from `corrcal.py` to a separate `wrapper.py` module.
- the C codes were renamed to `corrcal_c_funcs.c` and `corrcal_c_funcs.h`, and the output library was renamed to `c_corrcal*.so` to make them more clear.

piyanatk committed:
Use `np` namespace for numpy

piyanatk committed:
Use `np` namespace for numpy

ronniyjoseph commented on corrcal/test/corrcal_example.py:
```diff
@@ -1,33 +1,38 @@
-import numpy, time
-import corrcal
-from data import DATA_PATH
```
When I try and run the example, it fails because the data folder is in  a parent folder. Changing the import from data to ..data does not resolve the issue.

Hmm I remember us having issues with this

piyanatk replied:
This should be fixed in https://github.com/UKZN-Astronomy/corrcal/pull/21/commits/d357c3f38564de33b8d48da1eeacb977855ca70f.

The issue was that the data files were not included by `setuptools` when you run `pip install .`. Please uninstall `pip uninstall corrcal` and reinstall. Then, absolute import should work but make sure that you do not try to `import corrcal` from the local repository because it is usually at the top of the list in PYTHONPATH.

ronniyjoseph commented on corrcal/test/test_corrcal_fof.py:
```diff
@@ -1,67 +1,76 @@
-
-noise_org=numpy.ones(u_org.size)
-
-vis,u,v,noise,ant1,ant2,edges= corrcal.grid_data(vis_org, u_org, v_org, noise_org, ant1_org, ant2_org)
```
Traceback (most recent call last):
  File "test_corrcal_fof.py", line 37, in <module>
    vis,u,v,noise,ant1,ant2,edges= corrcal.grid_data(vis_org, u_org, v_org, noise_org, ant1_org, ant2_org)
ValueError: too many values to unpack (expected 7)

piyanatk replied:
Should be fixed in https://github.com/UKZN-Astronomy/corrcal/pull/21/commits/4c92827a0361dfc86daeb5e681e656fba8202a0a.

ronniyjoseph replied:
Issue is indeed fixed

ronniyjoseph commented on corrcal/corrcal.py:
```diff
@@ -1,141 +1,100 @@
            'mult_vecs_by_blocks', 'make_uv_from_antpos']
 
-
+# Not really sure what this does.
```
This allows you to use pyfof (a friends of friends implementation in python) to find redundant groups.
If pyfof is not installed the redundant grouping function "grid_data" cuts the uv-plane into a grid and selects baselines that fall into a grid cell as a redundant group.

See grid_data in wrapper.py

piyanatk replied:
Added these as a comment block in https://github.com/UKZN-Astronomy/corrcal/pull/21/commits/635a6eed3f47cf2b2155b6c109869b9f4f8ffc0a

ronniyjoseph commented on corrcal/corrcal.py:
```diff
@@ -1,141 +1,100 @@
 
 import numpy as np
 
+from .wrapper import *
```
We should get rid of this import *

piyanatk replied:
Fixed in https://github.com/UKZN-Astronomy/corrcal/pull/21/commits/635a6eed3f47cf2b2155b6c109869b9f4f8ffc0a

piyanatk committed:
Import a wrapper module instead of using a wildcard import.

- Wrapper functions were renamed accordingly.
- Some PEP8 formatting.

piyanatk committed:
Use setuptools-scm to automate versioning.

See https://pypi.org/project/setuptools-scm/.

piyanatk committed:
Make setuptools install `c_corrcal*.so` in the package directory.

piyanatk committed:
Specify `package_data` in `setup()`.

This makes sure that `*.dat files in `corrcal/data` are included as they are being used in the examples in `corrcal/test`.

piyanatk committed:
Fix unpacking bug.

I just added `ii` and `isconj` to the list of the return parameters to unpack as they are being returned from `grid_data`, but I have no idea what they are. They seem to not being used in the test file.

ronniyjoseph commented on corrcal/wrapper.py:
```diff
@@ -18,7 +18,7 @@
 # With ctypes, the location of libcorrcal*.so need to be hard-coded.
 # This will not be needed once we migrate to CFFI.
-LIB_LOC = glob(site.getsitepackages()[0] + "/c_corrcal*.so")[0]
+LIB_LOC = glob(site.getsitepackages()[0] + "/corrcal/c_corrcal*.so")[0]
```
I am getting issues with this import here.

Could you print the string stored under LIB_LOC.
Mine returns and empty list [].

piyanatk replied:
Hmm. It  is working fine on my machine. `LIB_LOC` can be access from `corrcal.wrapper.LIB_LOC`, and mine is pointing to the right file.

```python
In [10]: ls /Users/piyanat/src/miniconda3/envs/corrcal/lib/python3.7/site-packages/corrcal/                                                                                                          
__init__.py                      c_corrcal.cpython-37m-darwin.so* data/                            wrapper.py
__pycache__/                     corrcal.py                       test/

In [11]: import site                                                                                                                                                                                 

In [12]: site.getsitepackages()                                                                                                                                                                      
Out[12]: ['/Users/piyanat/src/miniconda3/envs/corrcal/lib/python3.7/site-packages']

In [13]: wrapper.LIB_LOC                                                                                                                                                                             
Out[13]: '/Users/piyanat/src/miniconda3/envs/corrcal/lib/python3.7/site-packages/corrcal/c_corrcal.cpython-37m-darwin.so'

```
 Can you:
- make sure that you have to most recent commit on restructure branch
- `pip uninstall corrcal` 
- reinstall it with `pip install .` from your local repository.
- launch `ipython` __*outside*__ of your local repository and try to import again.

ronniyjoseph replied:
Did it and problem is resolved, thanks!

steven-murray replied:
The problem of it loading the local package rather than the installed one is a common problem and the most suggested fix is to put your actual package in a (non-python-package) `src/` folder. So you'd have:

```
corrcal/
    setup.py
    src/
        corrcal/
            __init__.py
```

This will also help in testing (because pytest will by default pick up the local package rather than the installed one, which is _usually_ not what you want!)

ronniyjoseph replied:
Nice! Thanks for the suggestion @steven-murray

ronniyjoseph commented on corrcal/test/test_corrcal_fof.py:
```diff
@@ -1,67 +1,76 @@
+import numpy as np
+from scipy.optimize import fmin_cg
+
 from corrcal import corrcal
```
There seem to be issues with the wrapper module.
It seems that the corrcal c library is not where we think it is?

Traceback (most recent call last):
  File "test_corrcal_fof.py", line 4, in <module>
    from corrcal import corrcal
  File "/home/ronniyjoseph/Software/anaconda3/lib/python3.7/site-packages/corrcal-0.1.dev76+g4c92827-py3.7-linux-x86_64.egg/corrcal/__init__.py", line 1, in <module>
    from .corrcal import *
  File "/home/ronniyjoseph/Software/anaconda3/lib/python3.7/site-packages/corrcal-0.1.dev76+g4c92827-py3.7-linux-x86_64.egg/corrcal/corrcal.py", line 5, in <module>
    from . import wrapper
  File "/home/ronniyjoseph/Software/anaconda3/lib/python3.7/site-packages/corrcal-0.1.dev76+g4c92827-py3.7-linux-x86_64.egg/corrcal/wrapper.py", line 21, in <module>
    LIB_LOC = glob(site.getsitepackages()[0] + "/corrcal/c_corrcal*.so")[0]
IndexError: list index out of range

ronniyjoseph replied:
After uninstalling and reinstalling, issue has been resolved

ronniyjoseph reviewed (approved):
Version 0.1 works on my machine.
